log4j-api/src/main/java/org/apache/logging/log4j/internal/DefaultLogBuilder.java

```diff
@@ -0,0 +1,179 @@
+        this.inUse = false;
+        this.threadId = Thread.currentThread().getId();
+    }
+
+    public LogBuilder setLevel(Level level) {
+        this.inUse = true;
+        this.level = level;
+        this.marker = null;
+        this.throwable = null;
+        this.location = null;
+        this.object = null;
+        this.msg = null;
+        this.textMessage = null;
+        this.supplier = null;
+        this.parameters = null;
+        return this;
+    }
+
+    public LogBuilder withMarker(Marker marker) {
+        this.marker = marker;
+        return this;
+    }
+
+    public LogBuilder withThrowable(Throwable throwable) {
+        this.throwable = throwable;
+        return this;
+    }
+
+    public LogBuilder withLocation() {
+        location = StackLocatorUtil.getStackTraceElement(2);
+        return this;
+    }
+
+    public LogBuilder withLocation(StackTraceElement location) {
+        this.location = location;
+        return this;
+    }
+
+    public LogBuilder withMessage(String msg) {
+        this.textMessage = msg;
+        return this;
+    }
+
+    public LogBuilder withMessage(Object msg) {
+        this.object = msg;
+        return this;
+    }
+
+    public LogBuilder withMessage(Message msg) {
+        this.msg = msg;
+        return this;
+    }
+
+    public LogBuilder withMessage(Supplier<Message> supplier) {
+        this.supplier = supplier;
+        return this;
+    }
+
+    public LogBuilder withParameters(Object... params) {
+        if (params != null && params.length > 0) {
+            if (parameters == null) {
+                parameters = params;
+            } else {
+                Object[] prev = parameters;
+                int count = parameters.length + params.length;
+                parameters = new Object[count];
+                System.arraycopy(prev, 0, parameters, 0, prev.length);
+                System.arraycopy(params, 0, parameters, prev.length, params.length);
+            }
+        }
+        return this;
+    }
+
+     @SafeVarargs
+     public final LogBuilder withParameters(java.util.function.Supplier<Object>... params) {
+         if (params != null && params.length > 0) {
+             if (parameters == null) {
+                 parameters = new Object[params.length];
+                 for (int i = 0; i < params.length; ++i) {
+                    parameters[i] = params[i].get();
+                 }
+             } else {
+                 Object[] prev = parameters;
+                 int count = parameters.length + params.length;
+                 parameters = new Object[count];
+                 System.arraycopy(prev, 0, parameters, 0, prev.length);
+                 for (int i = 0; i < params.length; ++i) {
+                     parameters[prev.length + i] = params[i].get();
+                 }
+             }
+         }
+        return this;
+    }
+
+    public boolean isInUse() {
+        return inUse;
+    }
+
+    public void log() {
+        if (!inUse) {
```

Comment:
We can learn from [flogger](https://google.github.io/flogger/anatomy.html)'s approach.

Can we avoid the call to `log()` at the end? All that this does is signal that the log statement is done, like a call to `build()` at the end of a builder chain.

How about if we make the `log` method do more work? Something like this:

```java
logger.atTrace().log("Hi, {}", () -> fetchWorld());
logger.atTrace().withMarker(marker).withThrowable(t).log("Found error while {}", currentTask);
```

I propose we remove the `withMessage` and `withParameters` methods and replace them with overloaded `log` methods that take these parameters:

* Message
* Object
* Supplier<Message>
* String, Object...
* String, Supplier<Object>...
* (potentially other...)


Another benefit of this approach is that it would prevent users from doing the wrong thing, like for example:

```java
// the API should not allow this:
Message msg = msg();
logger.atTrace().withMessage(msg)
    .withMessage("It is now possible to use the API incorrectly")
    .withMessage(() -> differentMsg())
    .withParameters("a", "b", "c")
    .withParameters(() -> getParams())
    .withMessage("hi!")
    .log(); // now, which of the above should actually be logged?
```

Reply:
I should note that at Gary's request I removed "at" from all the logger methods. so the first would be
`logger.trace().log("Hi, {}", () -> fetchWordl());`

After reading the above a few times I think I am OK with this.

Reply:
About mixing Suppliers with other objects, this is still possible if (a) we cater for it in our implementation and (b) client code is willing to cast lambdas to `org.apache.logging.log4j.util.Supplier`. For example, this works fine in Java 8:

```java
static void log(String pattern, Object... something) {
    System.out.println(pattern);
    for (Object element : something) {
        if (element instanceof Supplier) {
            System.out.println(((Supplier) element).get());
        } else {
            System.out.println("Plain string: " + element);
        }
    }
}
public static void main(String[] args) {
    log("Hello {} {}", 123, (Supplier)() -> "I'm a little lambda");
}
```

Reply:
About the logger method naming: I'm generally in favor of short names and I liked Gary's suggestion. However, looking at it some more, in this case there may be something to be said for prefixing the new "fluent" methods with `at`;

* it makes the fluent methods stick out from the "normal" `info(String)`, `trace(Message)`, etc Logger  methods. I think it is nice if they look different, because they are quite different.
* users coding in an IDE will have the advantage of typing "at<Ctl+space>" and get completions for the fluent methods.
* it is nice if the `atLevel(Level)` method is consistently named with the `atInfo`, `atDebug`, etc other fluent methods.

@garydgregory (and others), what do you think?